sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java

```diff
@@ -52,280 +59,491 @@
+      @Override
+      public int hashCode() {
+        return uuid.hashCode();
+      }
+    }
   }
 
-  /** TestMatcher is a matcher designed for testing matcher serialization/deserialization. */
-  public static class TestMatcher extends BaseMatcher<PipelineResult>
-      implements SerializableMatcher<PipelineResult> {
-    private final UUID uuid = UUID.randomUUID();
+  /**
+   * Tests for {@link TestPipeline}'s detection of missing {@link Pipeline#run()}, or abandoned
+   * (dangling) {@link PAssert} or {@link org.apache.beam.sdk.transforms.PTransform} nodes.
+   */
+  @RunWith(Parameterized.class)
+  public static class TestPipelineEnforcementsTest implements Serializable {
 
-    @Override
-    public boolean matches(Object o) {
-      return true;
-    }
+    private static final List<String> WORDS = Collections.singletonList("hi there");
+    private static final String WHATEVER = "expected";
+    private static final String P_TRANSFORM = "PTransform";
+    private static final String P_ASSERT = "PAssert";
 
-    @Override
-    public void describeTo(Description description) {
-      description.appendText(String.format("%tL", new Date()));
-    }
+    private static class DummyRunner extends PipelineRunner<PipelineResult> {
 
-    @Override
-    public boolean equals(Object obj) {
-      if (!(obj instanceof TestMatcher)) {
-        return false;
+      @SuppressWarnings("unused") // used by reflection
+      public static DummyRunner fromOptions(final PipelineOptions opts) {
+        return new DummyRunner();
       }
-      TestMatcher other = (TestMatcher) obj;
-      return other.uuid.equals(uuid);
-    }
 
-    @Override
-    public int hashCode() {
-      return uuid.hashCode();
+      @Override
+      public PipelineResult run(final Pipeline pipeline) {
+        return new PipelineResult() {
+
+          @Override
+          public State getState() {
+            return null;
+          }
+
+          @Override
+          public State cancel() throws IOException {
+            return null;
+          }
+
+          @Override
+          public State waitUntilFinish(final Duration duration) {
+            return null;
+          }
+
+          @Override
+          public State waitUntilFinish() {
+            return null;
+          }
+
+          @Override
+          public <T> AggregatorValues<T> getAggregatorValues(final Aggregator<?, T> aggregator)
+              throws AggregatorRetrievalException {
+            return null;
+          }
+
+          @Override
+          public MetricResults metrics() {
+            return null;
+          }
+        };
+      }
     }
-  }
 
-  private static class DummyRunner extends PipelineRunner<PipelineResult> {
+    @SuppressWarnings("UnusedReturnValue")
+    private static PCollection<String> addTransform(final PCollection<String> pCollection) {
+      return pCollection.apply("Map2",
+                               MapElements.via(new SimpleFunction<String, String>() {
```

Comment:
FWIW we go with `google-java-format`. We can autoformat when merging, but it is most readable to just get that set up in your IDE or via command line.